src/dsa/rpo_falcon512/signature.rs

```diff
@@ -47,72 +47,71 @@ pub struct Signature {
     pub(super) pk: PublicKeyBytes,
     pub(super) sig: SignatureBytes,
 
-    // Cached polynomial decoding for public key and signatures
-    pub(super) pk_polynomial: OnceCell<Polynomial>,
-    pub(super) sig_polynomial: OnceCell<Polynomial>,
+    // Cached polynomial decoding for public key and signature
+    pub(super) pk_poly: Polynomial<FalconFelt>,
+    pub(super) sig_poly: Polynomial<FalconFelt>,
 }
 
 impl Signature {
     // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
     /// Returns the public key polynomial h.
-    pub fn pub_key_poly(&self) -> Polynomial {
-        *self.pk_polynomial.get_or_init(|| {
-            // we assume that the signature was constructed with a valid public key, and thus
-            // expect() is OK here.
-            Polynomial::from_pub_key(&self.pk).expect("invalid public key")
-        })
+    pub fn pub_key_poly(&self) -> &Polynomial<FalconFelt> {
+        &self.pk_poly
     }
 
     /// Returns the nonce component of the signature represented as field elements.
     ///
     /// Nonce bytes are converted to field elements by taking consecutive 5 byte chunks
     /// of the nonce and interpreting them as field elements.
-    pub fn nonce(&self) -> NonceElements {
+    pub fn nonce(&self) -> NonceBytes {
         // we assume that the signature was constructed with a valid signature, and thus
         // expect() is OK here.
-        let nonce = self.sig[1..41].try_into().expect("invalid signature");
-        decode_nonce(nonce)
+        self.sig[SIG_HEADER_LEN..SIG_HEADER_LEN + SIG_NONCE_LEN]
+            .try_into()
+            .expect("invalid signature")
     }
 
     // Returns the polynomial representation of the signature in Z_p[x]/(phi).
-    pub fn sig_poly(&self) -> Polynomial {
-        *self.sig_polynomial.get_or_init(|| {
-            // we assume that the signature was constructed with a valid signature, and thus
-            // expect() is OK here.
-            Polynomial::from_signature(&self.sig).expect("invalid signature")
-        })
+    pub fn sig_poly(&self) -> &Polynomial<FalconFelt> {
+        &self.sig_poly
     }
 
     // HASH-TO-POINT
     // --------------------------------------------------------------------------------------------
 
     /// Returns a polynomial in Z_p\[x\]/(phi) representing the hash of the provided message.
-    pub fn hash_to_point(&self, message: Word) -> Polynomial {
-        hash_to_point(message, &self.nonce())
+    pub fn hash_to_point(&self, message: Word) -> Polynomial<FalconFelt> {
+        let nonce = &self.nonce();
+        hash_to_point(message, nonce)
     }
```

Comment:
I think this is fine for now - though, I'll try to think of how to make this more generic.

Reply:
I think that we probably should change `message: Word` to `message: &[u8]`